### PR TITLE
add workloadRefPath to traitDefinition

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
Finalize the trait/workoad interaction design. We need to rely on the trait owner to signal if we can add the workloadRef field to its CR. It's very hard for appConfig controller to validate otherwise because

1. CRDs (such as all the existing OAM CRDs in crossplane) may not come with a strict validation schema
2. The trait can come with a strict validation webhook (i.e Knative route) that we have no way to know.

### CODE REVIEW

File: design/one-pager-trait-workload-interaction-mechanism.md [hunk 1]
```diff
@@ -122,9 +122,23 @@ The overall idea is for the applicationConfiguration controller to fill critical
 in the workload and trait CR it emits. In addition, we will provide a helper library so that
 trait controller developers can locate the resources they need with a simple function call. 
 Here is the list of changes that we propose.
-1. ApplicationConfig controller no longer assumes that all `trait` CRDs contain a "spec
+1. Add an optional field called `oamAware` to the `traitDefinition` schema. The trait owner needs
```
Comment (wonderflow) [inline design/one-pager-trait-workload-interaction-mechanism.md:125]: The name `oamAware` is not intuitive, what about "workloadDefNeeded"

File: design/one-pager-trait-workload-interaction-mechanism.md [hunk 2]
```diff
@@ -122,9 +122,23 @@ The overall idea is for the applicationConfiguration controller to fill critical
...
+         name: manualscalertraits.core.oam.dev
+       spec:
+         oamAware: true
+         definitionRef:
+           name: manualscalertraits.core.oam.dev
+     ```
+2. ApplicationConfig controller no longer assumes that all `trait` CRDs contain a "spec
 .workloadRef" field conforms to the OAM definition. It only fills the workload GVK to a `trait
-` CR if its CRD has a "spec.workloadRef" field defined as below.  
+` CR with "spec.workloadRef" field defined as below if the corresponding `traitDefiniton` is
```
Comment (zzxwill) [inline design/one-pager-trait-workload-interaction-mechanism.md:140]: +1

File: design/one-pager-trait-workload-interaction-mechanism.md (hunk 1 above)
Comment (ryanzhang-oss) [reply design/one-pager-trait-workload-interaction-mechanism.md:125]: I am open to naming. However, the reason for this name is we may reuse this field for future purposes, not limited to workload. So this is intended to be a general-purpose field.

Comment (resouer) [reply design/one-pager-trait-workload-interaction-mechanism.md:125]: Yeah, this is a bad name.

Actually, I would suggest change `workloadRef` to `targetRef`, and change this flag to `TargetAware`.

Note that `targetRef` also aligns with what Kubernetes is using today.

**Comment by hongchaodeng** [comment]
What would OAM runtime do if a trait signals "no"?

File: design/one-pager-trait-workload-interaction-mechanism.md (hunk 1 above)
Comment (ryanzhang-oss) [reply design/one-pager-trait-workload-interaction-mechanism.md:125]: @resouer can you point me to where `targetRef` is used in K8s?

**Comment by ryanzhang-oss** [comment]
> What would OAM runtime do if a trait signals "no"?

The default s "no" so there is no need for a trait to signal anything.  Also, the app Config controller will not patch the `workloadRef` field to a trait by default.

File: design/one-pager-trait-workload-interaction-mechanism.md (hunk 1 above)
Comment (resouer) [reply design/one-pager-trait-workload-interaction-mechanism.md:125]: Check this one: https://cloud.google.com/kubernetes-engine/docs/concepts/verticalpodautoscaler k8s has a built-in type for it.

**Comment by wonderflow** [comment]
@zzxwill  checked that dry-run=Server will work for checking if trait CR contains `workloadRef` field， which we don't need `oamAware` then.

**Comment by wonderflow** [comment]
LGTM

File: design/one-pager-trait-workload-interaction-mechanism.md [hunk 3]
```diff
@@ -179,10 +196,10 @@ deployment and service child resources.
 ## Impact to the existing system
 Here are the impacts of this mechanism to the existing OAM components
 - ApplicationConfiguration: This mechanism requires minimum changes in the
- applicationConfiguration controller.
+ applicationConfiguration controller except that it now needs to check if a trait is "oamAware".
```
Comment (hongchaodeng) [inline design/one-pager-trait-workload-interaction-mechanism.md:199]: > trait is "oamAware"
Please change this too

File: design/one-pager-trait-workload-interaction-mechanism.md [hunk 4]
```diff
@@ -122,9 +122,25 @@ The overall idea is for the applicationConfiguration controller to fill critical
 in the workload and trait CR it emits. In addition, we will provide a helper library so that
 trait controller developers can locate the resources they need with a simple function call. 
 Here is the list of changes that we propose.
-1. ApplicationConfig controller no longer assumes that all `trait` CRDs contain a "spec
+1. Add an optional field called `nativeTrait` to the `traitDefinition` schema. This is for the trait
```
Comment (hongchaodeng) [inline design/one-pager-trait-workload-interaction-mechanism.md:125]: nativeTrait is an ambiguous name.
It is very hard for a user to correlate "native" to the implementation of OAM trait/workload interaction mechanism.

File: design/one-pager-trait-workload-interaction-mechanism.md [hunk 5]
```diff
@@ -122,9 +122,25 @@ The overall idea is for the applicationConfiguration controller to fill critical
...
+       spec:
+         nativeTrait: true
+         definitionRef:
+           name: manualscalertraits.core.oam.dev
+     ```
+2. ApplicationConfig controller no longer assumes that all `trait` CRDs contain a "spec
 .workloadRef" field conforms to the OAM definition. It only fills the workload GVK to a `trait
-` CR if its CRD has a "spec.workloadRef" field defined as below.  
+` CR with "spec.workloadRef" field defined as below if the corresponding `traitDefiniton` is
+ marked as "oamAware".
```
Comment (hongchaodeng) [inline design/one-pager-trait-workload-interaction-mechanism.md:143]: This is "oamAware"?

File: design/one-pager-trait-workload-interaction-mechanism.md (hunk 3 above)
Comment (hongchaodeng) [inline design/one-pager-trait-workload-interaction-mechanism.md:199]: > trait is "oamAware"

Ditto. Needs to update.

File: design/one-pager-trait-workload-interaction-mechanism.md (hunk 1 above)
Comment (resouer) [reply design/one-pager-trait-workload-interaction-mechanism.md]: @ryanzhang-oss I will eat my word on the field name. `targetRef` is overlapping. Let's keep `workloadRef`.

File: design/one-pager-trait-workload-interaction-mechanism.md (hunk 4 above)
Comment (ryanzhang-oss) [reply design/one-pager-trait-workload-interaction-mechanism.md:125]: any suggestions?

**Pull request merged**